VACMS-2861: entity meta display block updates

acrollet opened the pull request:
## Description

See #2861 

## Testing done

Added behat test, manually tested.

## Screenshots

<img width="414" alt="Screen Shot 2020-10-21 at 3 21 45 PM" src="https://user-images.githubusercontent.com/101649/96791494-d47e7e00-13b5-11eb-8e1d-3b8f40a40654.png">

## QA steps

As a user with the content admin role:
1. Create a node without publishing it
1. [x] verify that the va.gov URL is not shown in the block
1. Publish the node
1. [x] verify that the va.gov URL is shown in the block, non-clickable with (Pending) appended
1. Archive the node
1. [x] verify that the va.gov URL is not shown in the block
1. Visit a node that has been previously published on production
1. [x] verify that the va.gov URL shows in the block and is clickable
1. Visit a revision page for a node
1. [x] verify that the block is shown

## Definition of Done
- [ ] Product release notes are written (or in progress), if required.
- [ ] Documentation has been updated, if applicable.

indytechcook commented on docroot/modules/custom/va_gov_workflow_assignments/src/Plugin/Block/EntityMetaDisplay.php:
```diff
@@ -75,28 +85,111 @@ public static function create(ContainerInterface $container, array $configuratio
+    $display_url = TRUE;
+
+    // Drupal sometimes hands us a nid and sometimes an upcasted node object.
+    // This code should be future-proof, cf:
```
```suggestion
    @TODO remove type checks when the patch is committed which looks to be 9.2.
    // This code should be future-proof, cf:
```

indytechcook commented on docroot/modules/custom/va_gov_workflow_assignments/src/Plugin/Block/EntityMetaDisplay.php:
```diff
@@ -75,28 +85,111 @@ public static function create(ContainerInterface $container, array $configuratio
+    elseif (is_numeric($this->routeMatch->getParameter('node'))) {
+      $node = \Drupal::entityTypeManager()->getStorage('node')->load($this->routeMatch->getParameter('node'));
+    }
+    if ($this->routeMatch->getParameter('node_revision') instanceof NodeInterface) {
```
```suggestion

    if ($this->routeMatch->getParameter('node_revision') instanceof NodeInterface) {
```

indytechcook commented on docroot/modules/custom/va_gov_workflow_assignments/src/Plugin/Block/EntityMetaDisplay.php:
```diff
@@ -75,28 +85,111 @@ public static function create(ContainerInterface $container, array $configuratio
       }
 
+      // The va.gov URL should be taken from the most recent published revision.
+      if ($display_url) {
```
Split this section out into it's own service name `VAGovStatus` or something.

indytechcook commented on docroot/modules/custom/va_gov_workflow_assignments/src/Plugin/Block/EntityMetaDisplay.php:
```diff
@@ -75,28 +85,111 @@ public static function create(ContainerInterface $container, array $configuratio
     $sections = $this->getSections()['links'];
     $block_items['Owner'] = $sections;
-    $nid = '';
+    $node = NULL;
```
Af first I had thought about spliting the logic to get a `node` and `node_revision` from a route out into it's own service but after more thought and the temporary nature of the code, let's put it into it's own method so it's easier to replace later.

indytechcook commented on docroot/modules/custom/va_gov_workflow_assignments/src/Plugin/Block/EntityMetaDisplay.php:
```diff
@@ -75,28 +85,111 @@ public static function create(ContainerInterface $container, array $configuratio
+      //
+      // First, get the latest published revision ID.
+      if ($display_url) {
+        $query = \Drupal::database()->select('content_moderation_state_field_revision', 'cmr')
```
Inject the database object instead.

indytechcook commented on docroot/modules/custom/va_gov_workflow_assignments/src/Plugin/Block/EntityMetaDisplay.php:
```diff
@@ -75,28 +85,111 @@ public static function create(ContainerInterface $container, array $configuratio
+      $node_revision = \Drupal::entityTypeManager()->getStorage('node')->loadRevision($this->routeMatch->getParameter('node_revision'));
+    }
+
+    if ($node) {
```
This is another section which could be broken into it's own service.

indytechcook commented on tests/behat/features/content_editing.feature:
```diff
@@ -97,3 +97,37 @@ Feature: CMS Users may effectively create & edit content
     Then I visit the "edit" page for a node with the title "Test Office - BeHaT"
     And the "menu[link_enabled]" checkbox should be checked
 
+@content_editing
```
YEAH FOR TESTS!!!  You are making the world a better place!

indytechcook commented on docroot/modules/custom/va_gov_workflow_assignments/src/Plugin/Block/EntityMetaDisplay.php:
```diff
@@ -75,28 +85,111 @@ public static function create(ContainerInterface $container, array $configuratio
+      // If there is an archived revision with an ID higher than the latest
+      // published revision ID, do not attempt to show the URL.
+      //
+      // First, get the latest published revision ID.
```
Comments should talk about why, let the code say what it's doing.

indytechcook commented on docroot/modules/custom/va_gov_workflow_assignments/src/Plugin/Block/EntityMetaDisplay.php:
```diff
@@ -75,28 +85,111 @@ public static function create(ContainerInterface $container, array $configuratio
+          $display_url = FALSE;
+        }
+
+        // Do not show the URL if there is an archived revision that is newer
```
Here is another "what" comment.  If you are adding comments like this to code, ask yourself how can I make the code speak for itself.

indytechcook reviewed (changes requested):
Overall the general logic is good.  The code needs to be organized into smaller services.

For example, I see the following services:

* `WorkflowAssignmentConfigService`: In charge of getting different configurations.
* `NodeWorkflowQueryManager`: In charge of finding information about the node/revision in relation to a workflow/content moderation.
* `VAGovService`: In charge of communicating with the published content site.

indytechcook commented on docroot/modules/custom/va_gov_workflow_assignments/src/Plugin/Block/EntityMetaDisplay.php:
```diff
@@ -75,28 +85,111 @@ public static function create(ContainerInterface $container, array $configuratio
+      // The va.gov URL should be taken from the most recent published revision.
+      if ($display_url) {
+        // Generate the va.gov URL.
+        $url = 'https://va.gov' . $node->toUrl()->toString();
```
Add some try/catch logic :)

```
The website encountered an unexpected error. Please try again later.

GuzzleHttp\Exception\ConnectException: cURL error 7: Failed to connect to va.gov port 443: Connection refused (see https://curl.haxx.se/libcurl/c/libcurl-errors.html) in GuzzleHttp\Handler\CurlFactory::createRejection() (line 200 of vendor/guzzlehttp/guzzle/src/Handler/CurlFactory.php).
```

acrollet commented:
@indytechcook I still have work to do on writing services for the revision moderation statuses and va.gov URL status checking, but if you have a moment to glance over the changes so far I'd appreciate it. I'm just wanting to make sure the direction is correct and get feedback on the structure and naming for the ExclusionTypes service. Thanks!

indytechcook commented on docroot/modules/custom/va_gov_backend/src/ExclusionTypes.php:
```diff
@@ -0,0 +1,70 @@
+  /**
+   * {@inheritDoc}
+   */
+  public static function create(ContainerInterface $container) {
```
You don't need this with a service.  The DI container will do it for you based upon what's definedi n the `servcies.yml` file.

indytechcook commented:
@acrollet this looks great so far.  The separation into different methods makes the code much easier read.  The service looks good.  I like that the service is specific to the exclude types rather than all config like I initially suggested.  The `typeIsExcluded` method is great.

I really like how this is progressing.  These two lines only make the code so much more readable

```
$node_revision = $this->getNodeRevision();
```

```
if ($this->vaGovUrlShouldBeDisplayed($node)) {
```

indytechcook commented on docroot/modules/custom/va_gov_backend/src/ExclusionTypesInterface.php:
```diff
@@ -0,0 +1,34 @@
+   * @return string
+   *   Json array.
+   */
+  public function getJson() : string;
```
What's this going todo? I didn't see it used yet.

acrollet replied:
@indytechcook this is a stub per @swirtSJW, cf. https://dsva.slack.com/archives/CT4GZBM8F/p1603742611304500?thread_ts=1603734185.284100&cid=CT4GZBM8F

acrollet replied:
@indytechcook thanks for the feedback!

indytechcook replied:
Very cool!  Thanks for the link

acrollet commented:
@indytechcook this is ready for another round of feedback :-) Also if you have a little time tomorrow (also @ndouglas if you're interested) I'd love to discuss how best to mock the responses for the VaGovUrl service in behat tests - I haven't come up with an approach I love.

acrollet commented:
ok @indytechcook @ndouglas @ElijahLynn you can see the approach I came up for mocking the va.gov responses during behat testing here: https://github.com/department-of-veterans-affairs/va.gov-cms/pull/3255/commits/3f9866ae436fdfb2e76611031760e24624d5284c

I don't completely love the approach, I would have preferred to keep the testing code within behat, but could not find an approach that would allow overriding a drupal service without creating a module. (Adding a services file using the container_yamls approach in settings.php would not work) Symfony has a neat way of doing [conditional arguments for services](https://symfony.com/doc/current/service_container/expression_language.html), but it's [not supported in Drupal](https://www.drupal.org/project/drupal/issues/2810587).

Feedback welcome :-)

indytechcook commented on docroot/modules/custom/va_gov_backend/modules/va_gov_backend_http_client/va_gov_backend_http_client.info.yml:
```diff
@@ -0,0 +1,7 @@
+core_version_requirement: ^8 || ^9
+package: 'Custom'
+dependencies:
+  - va_gov_backend
```
Add another line, this will fail phpcs.

indytechcook commented on docroot/modules/custom/va_gov_backend/modules/va_gov_backend_http_client/va_gov_backend_http_client.services.yml:
```diff
@@ -0,0 +1,4 @@
+services:
+  va_gov_backend_http_client.http_client:
+    decorates: 'http_client'
+    class: Drupal\va_gov_backend_http_client\VaGovBackendHttpClient
```
Add another line, this will fail phpcs.

indytechcook commented on docroot/modules/custom/va_gov_backend/src/Form/ExclusionTypesAdminForm.php:
```diff
@@ -51,16 +51,18 @@ public function buildForm(array $form, FormStateInterface $form_state) {
       '#description' => $this->t('Names of content types, that are not separate pages on va.gov'),
       '#weight' => '10',
-      '#default_value' => $config->get('types_to_exclude'),
+      '#default_value' => $config->get('types_to_exclude') ? $config->get('types_to_exclude') : [],
```
```suggestion
      '#default_value' => $config->get('types_to_exclude') ?? [],
```

acrollet replied:
TIL!

indytechcook commented on docroot/modules/custom/va_gov_backend/src/Service/VaGovUrl.php:
```diff
@@ -0,0 +1,85 @@
+  /**
+   * {@inheritDoc}
+   */
+  public function getVaGovUrlForEnvironment(String $environment) : string {
```
Allow this to be overridden by a `$settings` since the URL might be different for `demo` or `PR` environments.

acrollet replied:
@indytechcook can you expand on this a little? Are you thinking of something like this in a Drupal settings.php?

```
$settings['va_gov_web_environments'] = [
    'prod' => 'https://prod.demo.va.gov',
    'staging' => 'https://staging.demo.va.gov',
    'dev' => 'https://dev.demo.va.gov',
];
```

indytechcook replied:
No, just allow the url to be defined in `settings.php` directly.

acrollet replied:
@indytechcook done, let me know if this is what you had in mind?

indytechcook commented on docroot/modules/custom/va_gov_backend/src/Service/VaGovUrl.php:
```diff
@@ -0,0 +1,85 @@
+          return $e->getResponse->getStatusCode();
+        }
+
+        return 500;
```
Can you use `Symfony\Component\HttpFoundation\Response::HTTP_INTERNAL_SERVER_ERROR` instead?

acrollet replied:
I believe this is obviated by the new approach of returning success/failure, let me know if i missed it.

indytechcook commented on docroot/modules/custom/va_gov_workflow_assignments/src/Service/EditorialWorkflow.php:
```diff
@@ -0,0 +1,58 @@
+      ->orderBy('content_entity_revision_id', 'DESC')
+      ->range(0, 1);
+    $result = $query->execute()->fetchField();
+    return $result ? $result : 0;
```
```suggestion
    return $result ?? 0;
```

indytechcook commented on docroot/modules/custom/va_gov_workflow_assignments/src/Service/EditorialWorkflow.php:
```diff
@@ -0,0 +1,58 @@
+/**
+ * Class EditorialWorkflow.
+ */
+class EditorialWorkflow implements EditorialWorkflowInterface {
```
Let's name it `EditorialWorkflowContentRepository` or something.  `EditorialWorkflow` is a bit generic.

acrollet replied:
Sure, I was thinking that it could eventually be a more general service that could eventually be asked questions about the va.gov editorial workflow, but I'm good with keeping its purpose narrow.

indytechcook replied:
I understand that thought.  Those sorts of classes turn into giant blobs of unmanageable code :). I've done it too many times to count.

indytechcook commented on docroot/modules/custom/va_gov_workflow_assignments/src/Plugin/Block/EntityMetaDisplay.php:
```diff
@@ -187,11 +276,35 @@ private function getLink(Term $term, bool $first = FALSE) {
+   * @return bool
+   *   Boolean value.
+   */
+  private function vaGovUrlShouldBeDisplayed($node) {
```
```suggestion
  private function vaGovUrlShouldBeDisplayed($node) : bool {
```

indytechcook commented on docroot/modules/custom/va_gov_workflow_assignments/src/Plugin/Block/EntityMetaDisplay.php:
```diff
@@ -67,36 +104,46 @@ public static function create(ContainerInterface $container, array $configuratio
+    if ($this->vaGovUrlShouldBeDisplayed($node)) {
+      $va_gov_url = $this->vaGovUrl->getVaGovUrlForEntity($node);
+
+      if ($this->vaGovUrl->getVaGovUrlStatusForEntity($node) === 200) {
```
Let's have the servie determine success here instead of checking the status ourself

indytechcook reviewed (commented):
This is great!  I like how you mocked the test. It's a really good pattern for others to follow.

I made just a few comments.

ndouglas commented:
I think that's a fine solution, given the limitations.  I thought I'd done something along these lines at some point with creating a and injecting a service dynamically using a closure, but I can't find any relevant code in my repos or SO or anywhere else, so I must be mistaken.

indytechcook commented on docroot/modules/custom/va_gov_backend/src/Service/VaGovUrl.php:
```diff
@@ -0,0 +1,87 @@
+  /**
+   * {@inheritDoc}
+   */
+  public function vaGovUrlForEntityIsLive(EntityInterface $entity, String $environment = 'prod') : bool {
```
This is nice.

indytechcook commented on docroot/modules/custom/va_gov_backend/src/Service/VaGovUrl.php:
```diff
@@ -0,0 +1,87 @@
+      try {
+        // Keep the timeout low so that we don't block page loads for too long.
+        $response = $this->httpClient->head($va_gov_url, ['connect_timeout' => 2, 'http_errors' => FALSE]);
+        if ($response->getStatusCode() == 200) {
```
So I took a quick look at the `Symfony\Component\HttpFoundation\Response` object and there is a `isSuccessful` method!  

```
    /**
     * Is response successful?
     *
     * @return bool
     *
     * @final since version 3.2
     */
    public function isSuccessful()
    {
        return $this->statusCode >= 200 && $this->statusCode < 300;
    }
```

acrollet replied:
We're actually dealing with [GuzzleHttp\Psr7\Response](https://github.com/guzzle/psr7/blob/master/src/Response.php) rather than the Symfony class. However we can get similar behavior by using its default behavior (follow redirects and throw exceptions for 4xx/5xx responses) by returning TRUE if there is no exception.

indytechcook commented on docroot/modules/custom/va_gov_backend/modules/va_gov_backend_http_client/src/VaGovBackendHttpClient.php:
```diff
@@ -0,0 +1,26 @@
+   */
+  public function head($uri, array $options = []) {
+    if (preg_match('/-200$/', $uri)) {
+      return new Response(200);
```
Ideally, you would use `Response::create()` but it doesn't really matter especially in a test.

acrollet replied:
GuzzleHttp\Psr7\Response doesn't actually have a create method, but I'm glad you mentioned it bc I found an issue with the 404 response.

indytechcook commented:
Sorry to be so nit-picky here.   I see this as a template for other developers to use in their code.

indytechcook commented:
> I think that's a fine solution, given the limitations. I thought I'd done something along these lines at some point with creating a and injecting a service dynamically using a closure, but I can't find any relevant code in my repos or SO or anywhere else, so I must be mistaken.

It's easier in PHPUnit because you can override the DI container plus I think you can use annotation in the actually class to say which lass to use to mock it but I might be wrong there.

acrollet commented:
> Sorry to be so nit-picky here. I see this as a template for other developers to use in their code.

not at all! Thanks for taking the time.

acrollet commented:
@ElijahLynn @indytechcook @olivereri one thought I had is that we want to be really, really sure that the va_gov_backend_http_client is never enabled on production. Do we have anything running in the deploy process that enforces that the enabled modules match config?

ndouglas commented:
I've used [config_readonly](https://www.drupal.org/project/config_readonly) for a year or two.  Might be an option generally.

indytechcook commented:
We run `config:import` on deploy.  I like the `config_readonly` solution.  

I do think disabled `devel` and `ui` type modules during a deploy is a good idea.  Just incase someone enabled and forget to disable.

acrollet commented:
We may want a test that inspects the config for modules in a deny list and fails if it finds any, as it's a relatively easy developer mistake to update config and include the currently enabled modules.

ndouglas commented:
We could blacklist it in `config_split`.   I don't mind the idea of parsing the core.extensions.yml file and checking a blacklist though either.  Might prevent us from inadvertently enabling all sorts of potentially harmful or known-crappy modules.

indytechcook commented:
@ndouglas can you create a follow-up issue?

acrollet commented:
> **EDIT**: Oh man, if I could travel back in time and do this on my ex-employer's old D7 repo 😩

😂

acrollet commented:
@indytechcook tests passed, ready for final(?) review.

ElijahLynn commented on tests/behat/Drupal/ContextBase.php:
```diff
@@ -23,6 +23,22 @@ public function __construct() {
+   */
+  static function enableVaGovBackendHttpClient()
+  {
+    \Drupal::service('module_installer')->install(['va_gov_backend_http_client']);
```
This seems like it should work, but the install and uninstall seems like it may add 30-60 seconds onto the test run. Probably not an issue once we parallelize the tests though.

ElijahLynn commented:
> Do we have anything running in the deploy process that enforces that the enabled modules match config?

Yes, config split should work and can keep this module from ever being enabled on prod. I know Neil said to make a separate issue but I think it should be included in this PR if it is going to go in. 

https://github.com/department-of-veterans-affairs/va.gov-cms/tree/master/config

ndouglas commented:
> Yes, config split should work and can keep this module from ever being enabled on prod. I know Neil said to make a separate issue but I think it should be included in this PR if it is going to go in.

TBF Neil might've just been trying to get me to stop talking.

indytechcook commented on docroot/modules/custom/va_gov_backend/src/Service/VaGovUrl.php:
```diff
@@ -0,0 +1,100 @@
+ */
+class VaGovUrl implements VaGovUrlInterface {
+
+  const WEB_ENVIRONMENTS = [
```
We do not need `WEB_ENVIRONMENTS` here since we will be setting the URL in `settings.php`

indytechcook commented on docroot/modules/custom/va_gov_backend/src/Service/VaGovUrl.php:
```diff
@@ -0,0 +1,100 @@
+   */
+  public function getVaGovUrlForEnvironment(String $environment) : string {
+    // Allow prod URL to be overridden by settings.php.
+    if ($environment === 'prod' && $this->settings->get('va_gov_prod_url', '')) {
```
Only look at `va_gov_frontend_url` in `Settings` for this value.  Default to `https://prod.demo.va.gov`

indytechcook reviewed (commented):
Looking Great!  One _last_ change and we are good to go!

acrollet commented:
NB: this PR now introduces a minor visual change, which is that the carats between owner sections are no longer linked. I view this as an improvement, because:
* There can be ambiguity about which section link the carats are part of
* It's consistent with standard Drupal breadcrumbs below the tabs

Just calling it out in case it's an important requirement.

indytechcook commented on tests/phpunit/VaGovUrlServiceTest.php:
```diff
@@ -52,20 +52,20 @@ public function setUp() {
 
     $this->container = new ContainerBuilder();
-    $this->config = [];
+    $this->config = ['hash_salt' => 'SCVSPZNSKKK5XCRJ1WLE'];
```
What's the story to why this was added?

acrollet replied:
The test errored out locally without it. Not sure why it didn't before tbh.

indytechcook commented:
This all great!!

The last thing is to add the following to the settings.php files for the correct environment.

```
const WEB_ENVIRONMENTS = [
    'prod' => 'https://www.va.gov',
    'staging' => 'https://staging.va.gov',
    'dev' => 'https://dev.va.gov',
  ];
```

acrollet commented:
@indytechcook I must have misunderstood something in our previous conversation, I reverted the [BuildFrontend service](https://github.com/department-of-veterans-affairs/va.gov-cms/blob/5ee5496eb7ea028a152a3d21d6376b7f498aa7e5/docroot/modules/custom/va_gov_build_trigger/src/Service/BuildFrontend.php#L36) to its previous state. I already added a 16th minute item for today to discuss the testing blocker, should we discuss this then as well?

indytechcook commented:
@acrollet Tests are working great!  It's more for the block display on dev/stage/prod.  So adding `va_gov_frontend_url` to each will have the URL in the block on each environment display correctly.  This isn't a blocker though and only required if the front end URL is different between environments.

ElijahLynn commented:
> > Do we have anything running in the deploy process that enforces that the enabled modules match config?
> 
> Yes, config split should work and can keep this module from ever being enabled on prod. I know Neil said to make a separate issue but I think it should be included in this PR if it is going to go in.
> 
> https://github.com/department-of-veterans-affairs/va.gov-cms/tree/master/config

Discussed in Slack (internal) https://dsva.slack.com/archives/CT4GZBM8F/p1604359354131000?thread_ts=1604336649.101600&cid=CT4GZBM8F and the only way this would ever get enabled is if someone manually enables on PROD, or puts a config dependency on it so I am fine with not doing that now. I just thought it was a pretty low lift to config split it if it was going to go in.

ndouglas commented:
![maxresdefault](https://user-images.githubusercontent.com/1318579/98015558-89efff00-1dc2-11eb-9b31-e10250ec02d7.jpg)
